core/services/health/nurse.go

```diff
@@ -0,0 +1,382 @@
+package health
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"runtime/pprof"
+	"runtime/trace"
+	"sync"
+	"time"
+
+	"github.com/google/pprof/profile"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+type Nurse struct {
+	utils.StartStopOnce
+
+	cfg Config
+	log logger.Logger
+
+	checks   map[string]CheckFunc
+	checksMu sync.RWMutex
+
+	chGather chan gatherRequest
+	chStop   chan struct{}
+	wgDone   sync.WaitGroup
+}
+
+type Config struct {
+	ProfileRoot    string
+	PollInterval   time.Duration
+	GatherDuration time.Duration
+	MaxProfileSize utils.FileSize
+
+	CPUProfileRate       int
+	MemProfileRate       int
+	BlockProfileRate     int
+	MutexProfileFraction int
+
+	MemThreshold       utils.FileSize
+	GoroutineThreshold int
+}
+
+type CheckFunc func() (unwell bool, meta Meta)
+
+type gatherRequest struct {
+	reason string
+	meta   Meta
+}
+
+type Meta map[string]interface{}
+
+const profilePerms = 0666
+
+func NewNurse(cfg Config, log logger.Logger) *Nurse {
+	return &Nurse{
+		cfg:      cfg,
+		log:      log,
+		checks:   make(map[string]CheckFunc),
+		chGather: make(chan gatherRequest, 1),
+		chStop:   make(chan struct{}),
+	}
+}
+
+func (n *Nurse) Start() error {
+	return n.StartOnce("nurse", func() error {
+		// This must be set *once*, and it must occur as early as possible
+		runtime.MemProfileRate = n.cfg.MemProfileRate
```

Comment:
Would it be worthwhile to enforce that `Nurse` be a singleton?

Reply:
Considered it, but I was worried that it might interfere with our not-quite-e2e integration tests that spin up multiple nodes in the same process